presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1570,6 +1571,7 @@ private Scope computeAndAssignOrderByScope(OrderBy node, Scope sourceScope, Scop
             // ORDER BY should "see" both output and FROM fields during initial analysis and non-aggregation query planning
             Scope orderByScope = Scope.builder()
                     .withParent(sourceScope)
+                    .withScopeId(outputScope.getScopeId())
```

Comment:
> Remember: for identification purposes, only the effectively resolving scope matters.

I think it should be: for identification purposes, only the effectively resolving relation and it's field matters. That's the difference between `ScopeId` and `RelationId`. We care about field origin not the resolution mechanism (Scope).

> as is now: create scope with copy of output fields

Scope doesn't have fields. `RelationType` does. Scope only models parent-child dependency between relations.

> enhance scopes to support two parents and don't copy anything

Scope express simple parent-child relation. Sure, `RelationType` can have multiple parents if you compose a Scope out of it and connect it to a different parent/child.

In short: I don't think Scope and Relation are interchangeable concepts. One is merely a resolution mechanism with parent-child relationship and the other uniquely identifies fields and relations.